plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java

```diff
@@ -352,10 +345,13 @@ void updateRunnerButtons(MajorVersion majorVersion) {
   private void updateHierarchy(final MajorVersion majorVersion)
       throws InvocationTargetException, InterruptedException {
     // blocking call (regardless of "fork"), returning only after the inner runnable completes
-    getLaunchConfigurationDialog().run(true, true, new IRunnableWithProgress() {
+    getLaunchConfigurationDialog().run(true /*fork*/, true /*cancelable*/,
+        new IRunnableWithProgress() {
       @Override
       public void run(IProgressMonitor monitor)
           throws InvocationTargetException, InterruptedException {
+        monitor.beginTask(Messages.getString("loading.pipeline.options.hierarchy"), //$NON-NLS-1$
```

Comment:
This should be moved into `getPipelineOptionsHierarchy()` — `beginTask()` should be called by the actual worker.

Reply:
I just realized that `beginTask()` should be called only once on a given monitor instance, to mark the begging of a main task:
```
	 * Notifies that the main task is beginning.  This must only be called once
	 * on a given progress monitor instance.
```
So, it cannot be called it inside `getPipelineOptionsHierarchy()`. You could then guess `getPipelineOptionsHierarchy()` could do `subTask()` or `Submonitor.convert()`, but they don't work like `beginTask()`. (For example, `Submonitor.convert()` can't affect the amount work of the top-level dialog progress bar.)

I think it makes sense to `beginTask()` here to mark the beginning of a main task and specifically to declare the amount work of `IProgressMonitor.UNKNOWN` — it makes the top-level progress bar to swing left and right.

Reply:
If `getPipelineOptionsHierarchy()` want to do anything with the given `monitor`, it can always decided to do there with `SubMonitor`. Here, I think it's conceptually sane to begin a task on its own.

Reply:
It goes against the contract of `IProgressMonitor`.  Either pass through the monitor unmolested, OR you create a child monitor.

Why can't `getPipelineOptionsHierarchy()` be changed to:

```
  private PipelineOptionsHierarchy getPipelineOptionsHierarchy(IProgressMonitor monitor) {
    monitor.beginTask(xxx, IProgressMonitor.UNKNOWN);
    …
```

Reply:
> Why can't getPipelineOptionsHierarchy() be changed to:

What if the code is like this:

```java
public void run(IProgressMonitor monitor) throws InvocationTargetException, InterruptedException {
  getPipelineOptionsHierarchy(majorVersion, monitor);  // some dumb code
  anotherMethodThatCallsBeginTask(monitor);
  hierarchy = getPipelineOptionsHierarchy(majorVersion, monitor);
}
```
You end up calling `beginTask()` three times, whereas you should call it once. I think each method should not mess up with what their caller did and confine their scope to their own boundary, i.e., using `subTask` or `SubMonitor`.

Reply:
I think `beginTask()` is special, in that it marks the beginning of a task. You can call it if you are really beginning a task, even though you are not doing anything.

Reply:
In your example, you *must* pass in a child monitor of the original monitor.  Yes, each one calls `beginTask()`, but it's up to the *actual container that is reporting progress* to decide how sub-(sub-)*-monitors are reported.

(You can use `SubMonitor#convert()` to avoid needing a `beginTask()`.)

Reply:
Ah, I see. I got this wrong.

> (You can use SubMonitor#convert() to avoid needing a beginTask().)

As I said,

> (For example, SubMonitor.convert() can't affect the amount work of the top-level dialog progress bar.)

doing `convert()` doesn't give the result we want here. The progress bar won't swing. I'll make `getPipelineOptionsHierarchy()` call `beginTask()`.